circle.yml

```diff
@@ -1,6 +1,83 @@
+# These environment variables must be set in CircleCI UI
+#
+# DOCKERHUB_REPO - docker hub repo, format: <username>/<repo>
+# DOCKER_EMAIL   - login info for docker hub
+# DOCKER_USER
+# DOCKER_PASS
+
 machine:
+  pre:
+    - curl -sSL https://s3.amazonaws.com/circle-downloads/install-circleci-docker.sh | bash -s -- 1.10.0
   node:
     version: 4.3.1
+  services:
+    - docker
+
 dependencies:
+  cache_directories:
+    - "~/docker"
   pre:
     - npm install -g npm@3
+    - sudo apt-get update; sudo apt-get install pigz
+    - sudo pip install tox mozdownload mozinstall
+    - mozdownload --version latest --destination firefox.tar.bz2
+    - mozinstall firefox.tar.bz2
+    - wget -O geckodriver.tar.gz https://github.com/mozilla/geckodriver/releases/download/v0.10.0/geckodriver-v0.10.0-linux64.tar.gz
+    - gunzip -c geckodriver.tar.gz | tar xopf -
+    - chmod +x geckodriver && mv geckodriver /home/ubuntu/bin/
+
+  override:
+    - docker info
+
+    # Build the container, using Circle's Docker cache. Only use 1 image per
+    # day to keep the cache size down.
+    - I="image-$(date +%j).gz"; if [[ -e ~/docker/$I ]]; then echo "Loading $I"; pigz -d -c ~/docker/$I | docker load; fi
+
+    - docker build --pull -t $DOCKERHUB_REPO:amo --build-arg NODE_APP_INSTANCE=amo .
+    - docker build --pull -t $DOCKERHUB_REPO:disco --build-arg NODE_APP_INSTANCE=disco .
+    - docker build --pull -t $DOCKERHUB_REPO:admin --build-arg NODE_APP_INSTANCE=admin .
+
+    - docker images
+
+    # write the sha256 sum to an artifact to make image verification easier
+    - docker images --no-trunc | awk '/^app/ {print $3}' | tee $CIRCLE_ARTIFACTS/docker-image-shasum256.txt
+
+    # Clean up any old images; save the new one.
+    - I="image-$(date +%j).gz"; mkdir -p ~/docker; rm ~/docker/*; docker save ${DOCKERHUB_REPO} | pigz --fast -c > ~/docker/$I; ls -l ~/docker
+
+test:
+  override:
+    - docker run -d -p 4000:4000 -e NODE_APP_INSTANCE=disco -e NODE_ENV=uitests $DOCKERHUB_REPO:disco /bin/sh -c "npm run build && npm run start" && sleep 60
```

Comment:
This looks like a lot of custom configuration. Could we move this to something like a [docker-compose](https://docs.docker.com/compose/) yaml file? One thing we need to support is for developers to be able to easily reproduce the CI environment. In other words, if tests fail on CircleCI, a developer will need to build a docker image in exactly the same way to reproduce the failure.